lib/ScalarizePass.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018 The Clspv Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "clspv/Option.h"
+
+using namespace llvm;
+
+namespace {
+class ScalarizePass : public ModulePass {
+public:
+  static char ID;
+  ScalarizePass() : ModulePass(ID) {}
+
+  bool runOnModule(Module &M) override;
+
+private:
+  // Breaks a struct type phi down into its constituent elements. It does this
+  // recursively in the event the subtypes are also structs. Returns the
+  // replacement value. Returns the replacment value for the phi.
+  Value *ScalarizePhi(PHINode *phi);
+
+  // Phi nodes that need to be deleted.
+  std::vector<PHINode*> to_delete_;
+};
+} // namespace
+
+namespace clspv {
+ModulePass *createScalarizePass() { return new ScalarizePass(); }
+} // namespace clspv
+
+char ScalarizePass::ID = 0;
+static RegisterPass<ScalarizePass>
+    X("Scalarize", "Scalarizes some instructions with composite returns");
+
+bool ScalarizePass::runOnModule(Module &M) {
+  bool Changed = false;
+  for (auto &F : M) {
+    for (auto &BB : F) {
+      for (auto &I : BB) {
+        if (auto *phi = dyn_cast<PHINode>(&I)) {
+          if (clspv::Option::HackPhis() && phi->getType()->isStructTy()) {
```

Comment:
Until this pass is generalized, you could save time by pull this option check to the outside all the loops.  The if the option is off you don't traverse anything.